File: decompiled/RECIPE.1.8.4/be/business/connector/recipe/prescriber/services/RecipePrescriberServiceV4Impl.java
package be.business.connector.recipe.prescriber.services;

import be.business.connector.core.exceptions.IntegrationModuleException;
import be.business.connector.core.services.GenericWebserviceCaller;
import be.business.connector.projects.common.utils.EndpointResolver;
import be.fgov.ehealth.recipe.protocol.v4.CreatePrescriptionRequest;
import be.fgov.ehealth.recipe.protocol.v4.CreatePrescriptionResponse;
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionRequest;
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionResponse;
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionStatusRequest;
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionStatusResponse;
import be.fgov.ehealth.recipe.protocol.v4.GetValidationPropertiesRequest;
import be.fgov.ehealth.recipe.protocol.v4.GetValidationPropertiesResponse;
import be.fgov.ehealth.recipe.protocol.v4.ListFeedbacksRequest;
import be.fgov.ehealth.recipe.protocol.v4.ListFeedbacksResponse;
import be.fgov.ehealth.recipe.protocol.v4.ListOpenRidsRequest;
import be.fgov.ehealth.recipe.protocol.v4.ListOpenRidsResponse;
import be.fgov.ehealth.recipe.protocol.v4.ListRidsHistoryRequest;
import be.fgov.ehealth.recipe.protocol.v4.ListRidsHistoryResponse;
import be.fgov.ehealth.recipe.protocol.v4.PutFeedbackFlagRequest;
import be.fgov.ehealth.recipe.protocol.v4.PutFeedbackFlagResponse;
import be.fgov.ehealth.recipe.protocol.v4.PutVisionForPrescriberRequest;
import be.fgov.ehealth.recipe.protocol.v4.PutVisionForPrescriberResponse;
import be.fgov.ehealth.recipe.protocol.v4.RevokePrescriptionRequest;
import be.fgov.ehealth.recipe.protocol.v4.RevokePrescriptionResponse;
import be.fgov.ehealth.recipe.protocol.v4.SendNotificationRequest;
import be.fgov.ehealth.recipe.protocol.v4.SendNotificationResponse;

public class RecipePrescriberServiceV4Impl implements RecipePrescriberServiceV4 {
   private static final String ENDPOINT_NAME = "endpoint.prescriber.v4";
   private static RecipePrescriberServiceV4 recipePrescriberService;

   private RecipePrescriberServiceV4Impl() {
   }

   public static RecipePrescriberServiceV4 getInstance() {
      if (recipePrescriberService == null) {
         recipePrescriberService = new RecipePrescriberServiceV4Impl();
      }

      return recipePrescriberService;
   }

   public GetPrescriptionResponse getPrescriptionForPrescriber(GetPrescriptionRequest request) throws IntegrationModuleException {
      return (GetPrescriptionResponse)GenericWebserviceCaller.callGenericWebservice(request, GetPrescriptionResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:getPrescription\"");
   }

   public PutFeedbackFlagResponse putFeedbackFlag(PutFeedbackFlagRequest request) throws IntegrationModuleException {
      return (PutFeedbackFlagResponse)GenericWebserviceCaller.callGenericWebservice(request, PutFeedbackFlagResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:putFeedbackFlag\"");
   }

   public ListFeedbacksResponse listFeedbacks(ListFeedbacksRequest request) throws IntegrationModuleException {
      return (ListFeedbacksResponse)GenericWebserviceCaller.callGenericWebservice(request, ListFeedbacksResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:listFeedbacks\"");
   }

   public RevokePrescriptionResponse revokePrescription(RevokePrescriptionRequest request) throws IntegrationModuleException {
      return (RevokePrescriptionResponse)GenericWebserviceCaller.callGenericWebservice(request, RevokePrescriptionResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:revokePrescription\"");
   }

   public SendNotificationResponse sendNotification(SendNotificationRequest request) throws IntegrationModuleException {
      return (SendNotificationResponse)GenericWebserviceCaller.callGenericWebservice(request, SendNotificationResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:sendNotification\"");
   }

   public CreatePrescriptionResponse createPrescription(CreatePrescriptionRequest request) throws IntegrationModuleException {
      return (CreatePrescriptionResponse)GenericWebserviceCaller.callGenericWebservice(request, CreatePrescriptionResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:createPrescription\"");
   }

   public GetValidationPropertiesResponse getValidationProperties(GetValidationPropertiesRequest request) throws IntegrationModuleException {
      return (GetValidationPropertiesResponse)GenericWebserviceCaller.callGenericWebservice(request, GetValidationPropertiesResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:getValidationProperties\"");
   }

   public GetPrescriptionStatusResponse getPrescriptionStatus(GetPrescriptionStatusRequest request) throws IntegrationModuleException {
      return (GetPrescriptionStatusResponse)GenericWebserviceCaller.callGenericWebservice(request, GetPrescriptionStatusResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:getPrescriptionStatus\"");
   }

   public ListOpenRidsResponse listOpenRids(ListOpenRidsRequest request) throws IntegrationModuleException {
      return (ListOpenRidsResponse)GenericWebserviceCaller.callGenericWebservice(request, ListOpenRidsResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:ListOpenRids\"");
   }

   public PutVisionForPrescriberResponse putVisionForPrescriber(PutVisionForPrescriberRequest request) throws IntegrationModuleException {
      return (PutVisionForPrescriberResponse)GenericWebserviceCaller.callGenericWebservice(request, PutVisionForPrescriberResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:putVisionForPrescriber\"");
   }

   public ListRidsHistoryResponse listRidsHistory(ListRidsHistoryRequest request) throws IntegrationModuleException {
      return (ListRidsHistoryResponse)GenericWebserviceCaller.callGenericWebservice(request, ListRidsHistoryResponse.class, EndpointResolver.getEndpointUrlString("endpoint.prescriber.v4"), this.getClass().getName(), true, true, true, true, "\"urn:be:fgov:ehealth:recipe:protocol:v4:listRidsHistory\"");
   }
}

File: src/main/kotlin/org/taktik/connector/business/recipe/prescriber/services/RecipePrescriberServiceV4Impl.kt
package org.taktik.connector.business.recipe.prescriber.services

import be.fgov.ehealth.commons.protocol.v2.StatusResponseType
import be.fgov.ehealth.recipe.protocol.v4.CreatePrescriptionRequest
import be.fgov.ehealth.recipe.protocol.v4.CreatePrescriptionResponse
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionRequest
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionResponse
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionStatusRequest
import be.fgov.ehealth.recipe.protocol.v4.GetPrescriptionStatusResponse
import be.fgov.ehealth.recipe.protocol.v4.GetValidationPropertiesRequest
import be.fgov.ehealth.recipe.protocol.v4.GetValidationPropertiesResponse
import be.fgov.ehealth.recipe.protocol.v4.ListFeedbacksRequest
import be.fgov.ehealth.recipe.protocol.v4.ListFeedbacksResponse
import be.fgov.ehealth.recipe.protocol.v4.ListOpenRidsRequest
import be.fgov.ehealth.recipe.protocol.v4.ListOpenRidsResponse
import be.fgov.ehealth.recipe.protocol.v4.ListRidsHistoryRequest
import be.fgov.ehealth.recipe.protocol.v4.ListRidsHistoryResponse
import be.fgov.ehealth.recipe.protocol.v4.PutFeedbackFlagRequest
import be.fgov.ehealth.recipe.protocol.v4.PutFeedbackFlagResponse
import be.fgov.ehealth.recipe.protocol.v4.PutVisionForPrescriberRequest
import be.fgov.ehealth.recipe.protocol.v4.PutVisionForPrescriberResponse
import be.fgov.ehealth.recipe.protocol.v4.RequestType
import be.fgov.ehealth.recipe.protocol.v4.RevokePrescriptionRequest
import be.fgov.ehealth.recipe.protocol.v4.RevokePrescriptionResponse
import be.fgov.ehealth.recipe.protocol.v4.SendNotificationRequest
import be.fgov.ehealth.recipe.protocol.v4.SendNotificationResponse
import be.recipe.services.core.ResponseType
import be.recipe.services.prescriber.CreatePrescription
import org.taktik.connector.business.recipeprojects.common.utils.EndpointResolver
import org.taktik.connector.business.recipeprojects.core.exceptions.IntegrationModuleException
import org.taktik.connector.business.recipeprojects.core.services.GenericWebserviceCaller
import org.taktik.connector.technical.service.sts.security.Credential
import org.taktik.connector.technical.service.sts.security.SAMLToken


class RecipePrescriberServiceV4Impl : RecipePrescriberServiceV4 {
    private val ENDPOINT_NAME = "endpoint.recipe.prescriber.v4"

    @Throws(IntegrationModuleException::class)
    fun getPrescriptionForPrescriber(samlToken: SAMLToken, credential: Credential, request: GetPrescriptionRequest) =
        callGenericWs(samlToken, credential, request, GetPrescriptionResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:getPrescription\""
        )

    @Throws(IntegrationModuleException::class)
    fun putFeedbackFlag(samlToken: SAMLToken, credential: Credential, request: PutFeedbackFlagRequest) =
        callGenericWs(samlToken, credential, request, PutFeedbackFlagResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:putFeedbackFlag\""
        )

    @Throws(IntegrationModuleException::class)
    fun listFeedbacks(samlToken: SAMLToken, credential: Credential, request: ListFeedbacksRequest) =
        callGenericWs(samlToken, credential, request, ListFeedbacksResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:listFeedbacks\""
        )

    @Throws(IntegrationModuleException::class)
    fun revokePrescription(samlToken: SAMLToken, credential: Credential, request: RevokePrescriptionRequest) =
        callGenericWs(samlToken, credential, request, RevokePrescriptionResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:revokePrescription\""
        )

    @Throws(IntegrationModuleException::class)
    fun sendNotification(samlToken: SAMLToken, credential: Credential, request: SendNotificationRequest) =
        callGenericWs(samlToken, credential, request, SendNotificationResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:sendNotification\""
        )

    @Throws(IntegrationModuleException::class)
    fun createPrescription(samlToken: SAMLToken, credential: Credential, request: CreatePrescriptionRequest) =
        callGenericWs(samlToken, credential, request, CreatePrescriptionResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:createPrescription\""
        )

    @Throws(IntegrationModuleException::class)
    fun getValidationProperties(samlToken: SAMLToken, credential: Credential, request: GetValidationPropertiesRequest) =
        callGenericWs(samlToken, credential, request, GetValidationPropertiesResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:getValidationProperties\""
        )

    @Throws(IntegrationModuleException::class)
    fun getPrescriptionStatus(samlToken: SAMLToken, credential: Credential, request: GetPrescriptionStatusRequest) =
        callGenericWs(samlToken, credential, request, GetPrescriptionStatusResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:getPrescriptionStatus\""
        )

    @Throws(IntegrationModuleException::class)
    fun listOpenRids(samlToken: SAMLToken, credential: Credential, request: ListOpenRidsRequest) =
        callGenericWs(samlToken, credential, request, ListOpenRidsResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:ListOpenRids\""
        )

    @Throws(IntegrationModuleException::class)
    fun putVisionForPrescriber(samlToken: SAMLToken, credential: Credential, request: PutVisionForPrescriberRequest) =
        callGenericWs(samlToken, credential, request, PutVisionForPrescriberResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:putVisionForPrescriber\""
        )
    @Throws(IntegrationModuleException::class)
    fun listRidsHistory(samlToken: SAMLToken, credential: Credential, request: ListRidsHistoryRequest) =
        callGenericWs(samlToken, credential, request, ListRidsHistoryResponse::class.java,
            "\"urn:be:fgov:ehealth:recipe:protocol:v4:listRidsHistory\""
        )

    private fun <T: StatusResponseType> callGenericWs(
        samlToken: SAMLToken,
        credential: Credential,
        request: RequestType,
        clazz: Class<T>,
        method: String
    ): T {
        return GenericWebserviceCaller.callGenericWebservice(
            samlToken,
            credential,
            request,
            clazz,
            EndpointResolver.getEndpointUrlString(ENDPOINT_NAME),
            this.javaClass.name,
            true,
            true,
            true,
            true,
            method
        ) ?: throw IntegrationModuleException("error.connection.prescriber")
    }


}
